source/simulator/helper_functions.cc

```diff
@@ -909,8 +909,134 @@ namespace aspect
+                    compositional_values_12);
+                fe_values_21[introspection.extractors.compositional_fields[c]].get_function_values(vector,
+                    compositional_values_21);
+
+                local_compositional_average_integrals[c] = 0;
+
+                for (unsigned int q = 0; q < n_q_points_12; ++q)
+                  {
+                    local_compositional_average_integrals[c] += compositional_values_12[q]*fe_values_12.JxW(q);
+                  }
+                local_compositional_average_integrals[c] /= local_area_integral;
+
+                double min_composition=compositional_values_12[0];
+                double max_composition=compositional_values_12[0];
+                for (unsigned int q = 1; q < n_q_points_12; ++q)
+                  {
+                    min_composition = std::min<double> (min_composition,
+                                                        compositional_values_12[q]);
+                    max_composition = std::max<double> (max_composition,
+                                                        compositional_values_12[q]);
+                  }
+                for (unsigned int q = 0; q < n_q_points_21; ++q)
+                  {
+                    min_composition = std::min<double> (min_composition,
+                                                        compositional_values_21[q]);
+                    max_composition = std::max<double> (max_composition,
+                                                        compositional_values_21[q]);
+                  }
+                // find the trouble cell
+                if (min_composition <composition_exact_min || max_composition >composition_exact_max)
+                  {
+                    //Define theta: which is scaling constant
+                    double theta_T=std::min<double>(1,abs((composition_exact_max-local_compositional_average_integrals[c])
+                                                          /(max_composition-local_compositional_average_integrals[c])));
+                    theta_T=std::min<double>(theta_T,abs((composition_exact_min-local_compositional_average_integrals[c])
+                                                         /(min_composition-local_compositional_average_integrals[c])));
+                    //Modify the numerical solution at freedoms
+                    double t_tmp=0;
+                    for (unsigned int j = 0; j < finite_element.base_element(introspection.base_elements.compositional_fields).dofs_per_cell; ++j)
+                      {
+                        const unsigned int support_point_index
+                          = finite_element.component_to_system_index(introspection.component_indices.compositional_fields[c],
+                                                                     /*dof index within component=*/ j);
+                        t_tmp = distributed_vector(local_dof_indices[support_point_index]);
```

Comment:
At first I thought you were doing something iteratively here, because you reuse t_tmp in two consecutive lines. But what you really do is renaming one entry of the distributed vector and then using it for the calculation of the new value.
I would suggest to change this line to something like: `const double T_old = ...;` and the next one to: `const double T_limited = ...(T_old-...);`
I do not exactly know if this is faster or slower than your version, but it is definitely easier to understand.